#include <iostream>
#include <vector>
#include <string>
#include <thread>
#include <chrono>
#include <iomanip>
#include <sstream>
#include <memory>
#include <mutex>
#include <algorithm>
#include <cmath>

// å¼•å…¥ä½ çš„RPCå’ŒProtobufå¤´æ–‡ä»¶
#include "Krpcapplication.h"
#include "Krpcchannel.h"
#include "monitor.pb.h"

// FTXUI å¤´æ–‡ä»¶
#include <ftxui/dom/elements.hpp>
#include <ftxui/screen/screen.hpp>
#include <ftxui/component/component.hpp>
#include <ftxui/component/screen_interactive.hpp>
#include <ftxui/component/event.hpp>

using namespace ftxui;

// --- æ•°æ®ç»“æ„ ---
struct ServerMetrics {
    std::string name;
    float cpu_usage;
    float memory_usage;
    int64_t timestamp;
    bool online;
};

// --- å·¥å…·å‡½æ•° ---

// è·å–å½“å‰æ—¶é—´ HH:MM:SS
std::string GetCurrentTimeStr() {
    auto now = std::chrono::system_clock::now();
    auto time = std::chrono::system_clock::to_time_t(now);
    struct tm* timeinfo = localtime(&time);
    char buffer[20];
    strftime(buffer, sizeof(buffer), "%H:%M:%S", timeinfo);
    return std::string(buffer);
}

// æ ¼å¼åŒ–æ—¶é—´å·® (e.g. 5s ago)
std::string FormatElapsed(int64_t timestamp_ms) {
    if (timestamp_ms == 0) return "N/A";
    auto now = std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::system_clock::now().time_since_epoch()
    ).count();
    int64_t diff = (now - timestamp_ms) / 1000;
    if (diff < 0) diff = 0;
    return std::to_string(diff) + "s ago";
}

// ç»˜åˆ¶è¿›åº¦æ¡ï¼š[â–ˆâ–ˆâ–ˆâ–ˆâ–‘â–‘â–‘]
Element DrawCustomProgressBar(float percent) {
    percent = std::max(0.0f, std::min(100.0f, percent));
    int width = 20; // è¿›åº¦æ¡å†…éƒ¨å­—ç¬¦å®½åº¦
    int filled = static_cast<int>(std::round(width * (percent / 100.0f)));
    
    std::string bar_str;
    for (int i = 0; i < width; ++i) {
        if (i < filled) bar_str += "â–ˆ";
        else bar_str += "â–‘";
    }

    // é¢œè‰²é€»è¾‘
    Color c = Color::Green;
    if (percent > 80) c = Color::Red;
    else if (percent > 50) c = Color::Yellow;

    // ç»„åˆï¼šç™½è‰²æ‹¬å· + å½©è‰²æ¡
    return hbox({
        text("[") | color(Color::White),
        text(bar_str) | color(c),
        text("]") | color(Color::White)
    });
}

// --- ç›‘æ§ç³»ç»Ÿç±» ---

class DistributedMonitor {
public:
    DistributedMonitor() : selected_index_(0), show_details_(false) {
        // åˆå§‹åŒ– stub
        stub_ = std::make_unique<dmonitor::MonitorQueryServiceRpc_Stub>(new KrpcChannel(false));
    }

    void Run() {
        auto screen = ScreenInteractive::Fullscreen();

        // åå°åˆ·æ–°çº¿ç¨‹
        std::thread updater([this, &screen] {
            while (!should_exit_) {
                FetchData();
                screen.PostEvent(Event::Custom); // è§¦å‘é‡ç»˜
                std::this_thread::sleep_for(std::chrono::seconds(1));
            }
        });

        // æ¸²æŸ“é€»è¾‘è·¯ç”±
        auto component = Renderer([this] {
            if (show_details_) return RenderDetailsPage();
            return RenderMainPage();
        });

        // äº‹ä»¶å¤„ç†
        component = CatchEvent(component, [this, &screen](Event event) {
            if (show_details_) {
                if (event == Event::Escape || event == Event::Character('q') || event == Event::Backspace) {
                    show_details_ = false;
                    return true;
                }
            } else {
                if (event == Event::ArrowDown || event == Event::Character('j')) {
                    if (!servers_.empty()) selected_index_ = (selected_index_ + 1) % servers_.size();
                    return true;
                }
                if (event == Event::ArrowUp || event == Event::Character('k')) {
                    if (!servers_.empty()) selected_index_ = (selected_index_ + servers_.size() - 1) % servers_.size();
                    return true;
                }
                if (event == Event::Return) {
                    if (!servers_.empty()) show_details_ = true;
                    return true;
                }
                if (event == Event::Escape || event == Event::Character('q')) {
                    screen.ExitLoopClosure()();
                    should_exit_ = true;
                    return true;
                }
            }
            return false;
        });

        screen.Loop(component);
        should_exit_ = true;
        if (updater.joinable()) updater.join();
    }

private:
    std::unique_ptr<dmonitor::MonitorQueryServiceRpc_Stub> stub_;
    std::vector<ServerMetrics> servers_;
    std::mutex data_mutex_;
    bool should_exit_ = false;
    int selected_index_;
    bool show_details_;

    // è·å–æ•°æ®
    void FetchData() {
        dmonitor::QueryRequest req;
        dmonitor::QueryResponse rsp;
        req.set_server_name(""); // ç©ºå­—ç¬¦ä¸²è¡¨ç¤ºè·å–æ‰€æœ‰
        stub_->Query(nullptr, &req, &rsp, nullptr);

        std::lock_guard<std::mutex> lock(data_mutex_);
        if (rsp.metrics_size() > 0) {
            servers_.clear();
            for (int i = 0; i < rsp.metrics_size(); i++) {
                const auto& m = rsp.metrics(i);
                servers_.push_back({
                    m.server_name(),
                    m.cpu_usage(),
                    m.memory_usage(),
                    m.timestamp(),
                    m.cpu_usage() >= 0
                });
            }
            // é˜²æ­¢ç´¢å¼•è¶Šç•Œ
            if (selected_index_ >= servers_.size()) selected_index_ = 0;
        }
    }

    // ä¸»é¡µé¢æ¸²æŸ“
    Element RenderMainPage() {
        std::lock_guard<std::mutex> lock(data_mutex_);

        // 1. æ ‡é¢˜è¡Œ
        auto title_row = hbox({
            text(" DISTRIBUTED SYSTEM MONITOR ") | bold | flex,
            text(" " + GetCurrentTimeStr() + " ") | bold 
        }) | color(Color::BlueLight);

        // å®šä¹‰åˆ—å®½ (è€ƒè™‘åˆ°æ‹¬å·å¢åŠ äº†å®½åº¦ï¼Œé€‚å½“è°ƒå®½)
        const int w_arrow = 2;
        const int w_name = 20;
        const int w_status = 15;  // å¢åŠ å®½åº¦ä»¥å®¹çº³emoji
        const int w_cpu = 35; 
        const int w_mem = 35; 

        // 2. è¡¨å¤´ (å»æ‰ç«–çº¿ï¼Œåªä¿ç•™ç©ºæ ¼é—´è·)
        auto header_row = hbox({
            text("  ")            | size(WIDTH, EQUAL, w_arrow), // å ä½ç»™ç®­å¤´
            text("SERVER NAME")   | size(WIDTH, EQUAL, w_name) | bold,
            text("     "), 
            text("STATUS")        | size(WIDTH, EQUAL, w_status) | bold,
            text("   "), 
            text("CPU USAGE")     | size(WIDTH, EQUAL, w_cpu) | bold,
            text("   "), 
            text("MEM USAGE")     | size(WIDTH, EQUAL, w_mem) | bold
        });

        // 3. æ•°æ®è¡Œ
        Elements rows;
        if (servers_.empty()) {
            rows.push_back(text("Waiting for data...") | center | flex);
        } else {
            for (size_t i = 0; i < servers_.size(); ++i) {
                const auto& svr = servers_[i];
                bool is_selected = (i == selected_index_);

                // é€‰ä¸­æŒ‡ç¤ºç¬¦ (ä½¿ç”¨emojiç®­å¤´)
                std::string prefix = is_selected ? "> " : "  "; // â–¶ æˆ– > æˆ–
                
                // åœ¨çº¿çŠ¶æ€ï¼ˆä½¿ç”¨æ›´å¤§çš„emojiå›¾æ ‡ï¼‰
                auto status = svr.online 
                    ? text("ğŸŸ¢ ONLINE") | color(Color::Green) 
                    : text("ğŸ”´ OFFLINE") | color(Color::Red);

                // è¿›åº¦æ¡ç»„åˆï¼ˆå¤„ç†è´Ÿæ•°æ˜¾ç¤ºä¸º0ï¼‰
                int cpu_percent = std::max(0, (int)svr.cpu_usage);
                int mem_percent = std::max(0, (int)svr.memory_usage);
                
                auto cpu_bar = hbox({
                    DrawCustomProgressBar(cpu_percent),
                    text(" " + std::to_string(cpu_percent) + "%")
                });
                
                auto mem_bar = hbox({
                    DrawCustomProgressBar(mem_percent),
                    text(" " + std::to_string(mem_percent) + "%")
                });

                // æ„å»ºå•è¡Œ (æ— èƒŒæ™¯è‰²ï¼Œæ— ç«–çº¿)
                auto row = hbox({
                    text(prefix)   | size(WIDTH, EQUAL, w_arrow) | color(Color::BlueLight) | bold,
                    text(svr.name) | size(WIDTH, EQUAL, w_name) | (is_selected ? (color(Color::Blue) | bold) : color(Color::White)),
                    text("   "),
                    status         | size(WIDTH, EQUAL, w_status),
                    text("   "),
                    cpu_bar        | size(WIDTH, EQUAL, w_cpu),
                    text("   "),
                    mem_bar        | size(WIDTH, EQUAL, w_mem)
                });

                rows.push_back(row);
                // ã€å…³é”®ç‚¹ã€‘æ¯è¡Œæ•°æ®åå¢åŠ ä¸€ä¸ªç©ºè¡Œï¼Œå¢åŠ å‚ç›´é—´è·
                rows.push_back(text("")); 
            }
        }

        // 4. åº•éƒ¨æ  (ä½¿ç”¨ç®­å¤´ç¬¦å·)
        auto footer_row = text(" [â†‘/â†“] Select   [Enter] Details   [q/Esc] Quit ") | center | dim;

        // 5. æ•´ä½“ç»„è£…
        return vbox({
            title_row,
            separator(), 
            vbox({
                header_row,
                separator(), // è¡¨å¤´ä¸‹æ–¹æ·»åŠ åˆ†éš”çº¿
                // text(""), // è¡¨å¤´å’Œå†…å®¹ä¹‹é—´ç©ºä¸€è¡Œ
                // ä½¿ç”¨ margin(1) åœ¨è¾¹æ¡†å†…éƒ¨åˆ¶é€ å†…è¾¹è·
                vbox(std::move(rows)) | yframe | borderEmpty
            }) | flex, 
            separator(),
            footer_row
        }) | border;
    }

    // è¯¦æƒ…é¡µæ¸²æŸ“
    Element RenderDetailsPage() {
        std::lock_guard<std::mutex> lock(data_mutex_);
        if (servers_.empty() || selected_index_ >= servers_.size()) {
            return text("Error: No server selected");
        }
        
        const auto& svr = servers_[selected_index_];

        // å¤„ç†è´Ÿæ•°æ˜¾ç¤ºä¸º0
        int cpu_percent = std::max(0, (int)svr.cpu_usage);
        int mem_percent = std::max(0, (int)svr.memory_usage);

        auto content = vbox({
            text(" Server Details: " + svr.name) | bold,
            separator(),
            hbox({ text("Status: "), svr.online ? text("ğŸŸ¢ ONLINE") | color(Color::Green) : text("ğŸ”´ OFFLINE") | color(Color::Red) }),
            text(""),
            text("Last Heartbeat:     " + FormatElapsed(svr.timestamp)),
            text("IP Address:         192.168.1.X (Placeholder)"),
            text(""),
            text("Realtime Metrics:") | bold,
            hbox({ text("CPU: "), DrawCustomProgressBar(cpu_percent) | flex, text(" " + std::to_string(cpu_percent) + "%") }),
            hbox({ text("MEM: "), DrawCustomProgressBar(mem_percent) | flex, text(" " + std::to_string(mem_percent) + "%") }),
        }) | border | size(WIDTH, GREATER_THAN, 60) | center;

        return vbox({
            filler(),
            content,
            filler(),
            text("Press [Esc] to return") | center
        });
    }
};

int main(int argc, char* argv[]) {
    // åˆå§‹åŒ– RPC
    KrpcApplication::Init(argc, argv);

    // è¿è¡Œ TUI
    DistributedMonitor monitor;
    monitor.Run();

    return 0;
}